temporal-sdk/src/main/java/io/temporal/common/converter/SearchAttributesPayloadConverter.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.common.converter;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.protobuf.ByteString;
+import io.temporal.api.common.v1.Payload;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class SearchAttributesPayloadConverter {
+
+  private static ObjectMapper mapper;
+  private static final Logger log = LoggerFactory.getLogger(SearchAttributesPayloadConverter.class);
+  public static final SearchAttributesPayloadConverter INSTANCE =
+      new SearchAttributesPayloadConverter();
+
+  private SearchAttributesPayloadConverter() {
+    mapper = new ObjectMapper();
+    mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+    mapper.registerModule(new JavaTimeModule());
+    mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+  }
+
+  public Optional<Payload> toData(Object obj) throws DataConverterException {
+    try {
+      byte[] serialized;
+      if (obj instanceof LocalDateTime) {
+        ZonedDateTime utc = ((LocalDateTime) obj).atZone(ZoneOffset.UTC);
+        serialized = mapper.writeValueAsBytes(utc);
+      } else {
+        serialized = mapper.writeValueAsBytes(obj);
+      }
+      String type = SearchAttributesUtil.javaTypeToEncodedType(obj.getClass()).name();
+      return Optional.of(
+          Payload.newBuilder()
+              .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, EncodingKeys.METADATA_ENCODING_JSON)
+              .putMetadata(EncodingKeys.METADATA_TYPE_KEY, ByteString.copyFromUtf8(type))
+              .setData(ByteString.copyFrom(serialized))
+              .build());
+    } catch (JsonProcessingException e) {
+      throw new DataConverterException(e);
+    }
+  }
+
+  public Object fromData(Payload payload) throws DataConverterException {
+    ByteString data = payload.getData();
+    ByteString type = payload.getMetadataMap().get(EncodingKeys.METADATA_TYPE_KEY);
```

Comment:
Per our discussion, please provide a separate code path to deserialize old-style system search attributes in case the server doesn't return `type` field back.

Reply:
But it does! 😃

Reply:
It doesn't and I verified it with @alexshtin 
You got this behavior because you now send type field to the server when you save and old servers were saving this metadata and reading it back. That's how you got "type" field back.
If you will run this code against an old server with an old workflow initiated by old JavaSDK - you will break backward compatibility.
Feel free to ask @alexshtin for more details, he has the full context.